rbd-mirror: delayed replication support

trociny opened the pull request:
Fixes: http://tracker.ceph.com/issues/15371
Signed-off-by: Mykola Golub <mgolub@mirantis.com>

dillaman commented:
Post-Kraken feature

trociny commented on src/librbd/journal/Types.cc:
```diff
@@ -384,11 +385,15 @@ void EventEntry::decode(bufferlist::iterator& it) {
 
   boost::apply_visitor(DecodeVisitor(struct_v, it), event);
+  if (struct_v >= 4) {
+    ::decode(timestamp, it);
```
@dillaman Thinking about this a little more, not sure it will always work. If a new event type is added (or an existent one is extended) I expect it will fail.

Actually, I would like timestamp be encoded at the beginning of the EventEntry, before Event, but I don't see how to preserve compatibility, so an older client still could read a journal created by a new client. A way I see is to 
make timestamp a member of Event, i.e. adding to every event structure (or use inheritance). I am not like this much though.

Do you have any suggestions? Do we want older clients to be able to decode newer journals?

dillaman replied:
New fields always need to be added to the end of the structure to preserve compatibility.  You are correct that by placing the encode/decode in this common location, there is the possibility that an updated ```Event``` will now be creating incompatible streams. I think you need to add this to every event structure.

dillaman commented on src/common/config_opts.h:
```diff
@@ -1272,6 +1272,7 @@ OPTION(rbd_validate_pool, OPT_BOOL, true) // true if empty pools should be valid
 OPTION(rbd_validate_names, OPT_BOOL, true) // true if image specs should be validated
 OPTION(rbd_auto_exclusive_lock_until_manual_request, OPT_BOOL, true) // whether to automatically acquire/release exclusive lock until it is explicitly requested, i.e. before we know the user of librbd is properly using the lock API
 OPTION(rbd_mirroring_resync_after_disconnect, OPT_BOOL, false) // automatically start image resync after mirroring is disconnected due to being laggy
+OPTION(rbd_mirroring_delay, OPT_INT, 0) // time-delay in seconds for rbd-mirror asynchronous replication
```
Minor: perhaps ```rbd_mirroring_replication/replay_delay```?

dillaman commented on src/tools/rbd_mirror/ImageReplayer.cc:
```diff
@@ -225,6 +225,19 @@ class ImageReplayerAdminSocketHook : public AdminSocketHook {
+      return 0;
+    }
+
+    return (event_time + mirroring_delay - now) + 1;
```
Minor: why add an extra second?

trociny replied:
It is to convert a double to the smallest following int (I could use ceil() but thought my version was good enough for this purposes).

dillaman commented on src/tools/rbd_mirror/ImageReplayer.cc:
```diff
@@ -225,6 +225,19 @@ class ImageReplayerAdminSocketHook : public AdminSocketHook {
+      return 0;
+    }
+
+    utime_t now = ceph_clock_now(g_ceph_context);
```
Minor: I think we want a new time helper to retrieving the UTC time, otherwise it would rely on users correctly setting the timezone offset and keeping it updated w/ daylight savings, etc.

trociny replied:
As I see, ceph_clock_now() uses clock_gettime(), which gets time since the Epoch, so it should not depend on the user timezone?

Also, not sure about using g_ceph_context here. In journal/Types.h I used nullptr, just not to add an additional header file from common. I think this should be consistent, i.e. nullptr here too. On the other hand, being able to shift time via clock_offset config option could be useful for testing, so may be use g_ceph_context here and when encoding?

dillaman replied:
Well, if you are using UTC there should be no need to shift.

dillaman commented on src/tools/rbd_mirror/ImageReplayer.cc:
```diff
@@ -225,6 +225,19 @@ class ImageReplayerAdminSocketHook : public AdminSocketHook {
   Commands commands;
 };
 
+int replay_delay(const utime_t &event_time, int mirroring_delay) {
```
Minor: perhaps ```uint32_t calculate_replay_delay```?

dillaman commented:
I would like to see a mock unit test for the delay where it properly handles a stop while waiting for the delay to complete.

dachary commented:
jenkins test this please (no log)

trociny commented:
@dillaman I am still working on ImageReplayer mock tests. Your other comments are addressed and I would like to know what do you think about my approach for encoding the timestamp? I wouldn't like to add the timestamp into every event entry type. I think, by encoding it in an additional ENCODE_START/FINISH section, I have addressed potential compatibility issues, not sure about the interface (EventMetada type) though, how nice it looks...

dillaman commented:
@trociny I like the ```EventMetadata``` wrapper -- but why not just add it to ```EventEntry``` and have its ```encode``` / ```decode``` methods encode/decode the metadata after encoding/decoding the actual event?

trociny commented:
@dillaman I though having a dedicated object for every ENCODE_START/FINISH section would be a little cleaner, but I don't have a strong opinion here, so I will do as you suggest.

dillaman commented:
@trociny My rationale is just because ```EventEntry``` is already a wrapper -- it runs the encode/decode on the actual ```Event``` type it contains. Therefore, it seems like a natural extension point instead of adding external helpers for encoding/decoding the variant and the metadata record.

trociny commented:
@dillaman Then do we really need a new struct EventMetadata? May be just add timestamp to EventEntry, end have it something like below?
```
void EventEntry::encode(bufferlist& bl) const {
  ENCODE_START(3, 1, bl);
  boost::apply_visitor(EncodeVisitor(bl), event);
  ENCODE_FINISH(bl);
  ENCODE_START(1, 1, bl);
  ::encode(timestamp, bl);
  ENCODE_FINISH(bl);
}
```

Or do you think EventMetadata still could be useful?

dillaman commented:
@trociny That approach is fine with me as well.

dachary commented:
jenkins test this please (jenkins failure)

trociny commented:
@dillaman updated (timestamp encoding, and ImageReplayer mock tests added)

dillaman commented on src/librbd/Journal.cc:
```diff
@@ -710,7 +710,8 @@ int Journal<I>::demote() {
 
-    journal::EventEntry event_entry{journal::DemoteEvent{}};
+    journal::EventEntry event_entry{journal::DemoteEvent{},
+                                    ceph_clock_now(nullptr)};
```
My mistake -- I didn't realize they tracked from the Epoch.

trociny commented:
tests updated after the recent changes to master

dillaman commented:
@trociny Can you rebase this to the latest master?  It looks like ```ceph_clock_now``` no longer takes an argument.

https://github.com/ceph/ceph/blob/master/src/common/Clock.h#L22

trociny commented:
@dillaman updated

dillaman commented:
@trociny The build is still failing on Xenial [1] (works on Trusty and CentOS) 

[1] https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=xenial,DIST=xenial,MACHINE_SIZE=huge/317/console

trociny commented:
@dillaman I fixed this by adding a dummy ImageSync class to test_mock_ImageReplayer.cc.

dillaman commented:
@trociny thanks

dillaman commented:
@trociny Looks like ceph_test_rbd_mirror fails w/ "Expected: (delay) >= (DELAY), actual: 2.09607 vs 10"

http://qa-proxy.ceph.com/teuthology/jdillaman-2017-01-06_11:21:00-rbd-wip-jd-testing-distro-basic-smithi/695591/teuthology.log
http://qa-proxy.ceph.com/teuthology/jdillaman-2017-01-06_11:21:00-rbd-wip-jd-testing-distro-basic-smithi/695685/teuthology.log

trociny commented:
@dillaman It turned out I didn't test well it after the latest rebase , after metadata replication had been merged. I just checked unittest and it worked. My test used set_metadata to set delay param on local image, and after the rebase I had to update the test to set metadata param on the remote image and wait for it to be replicated.

I thought after metadata_set operation the corresponding ImageCtx parameter would be updated automatically (by refresh request) but it turned out it is only updated in open request.

I changed the test to  set the delay using `m_local_cluster->conf_set` instead of set_metadata, so it should work now.

But don't you think it is a bug that metadata is not updated by refresh request?

Also, interesting, I investigated why the old version worked with librados test stub. It turned out, that this test would pass on the librados test stub and fail on a real rados:
```
  open_remote_image(&ictx);
  ictx->cct->_conf->set_val("rbd_mirroring_replay_delay", "111");
  close_image(ictx);
  open_local_image(&ictx);
  ASSERT_EQ(111, ictx->cct->_conf->rbd_mirroring_replay_delay);
  close_image(ictx);
```
So, it looks like in the librados test stub config is shared between remote and local contexts, while in the real rados it is not. metadata_set operation also changes the current config value, so in the old test version, when metdata_set was executed for the remote context it also updated the config value for local context and the test worked.

Actually, I am also not sure that this is correct to change config value after metadata_set operation. I would rather expect the corresponding ImgCtx variable is only updated.

dillaman commented:
@trociny I don't necessarily think it's a bug that the config overrides aren't dynamically picked up via metadata set since no other rbd config values are dynamic (via asok config set).

The issue w/ librados_test_stub is probably due to the fact that it only really supports a singleton version of RadosClient. It's the same reason why blacklisting cannot work w/o a slight refactor. I guess it's never risen to the level of importance to tackle yet, though.

dillaman commented:
retest this please

trociny commented:
> I don't necessarily think it's a bug that the config overrides aren't dynamically picked up via metadata set since no other rbd config values are dynamic (via asok config set).

I can agree that it might be not necessary to update relevant ImageCtx vars on metadata_set operation. But still it looks strange that we update contex config variables  on metadata update operation. That means if a user executes metadata_set for one image, and then open another image it will use metadata settings for the previous image in this session.

dillaman commented:
@trociny Oh -- I see what you are saying now. The fact that ```Operations<I>::metadata_set``` is setting the configuration setting is a bug. It should use a dummy ```md_config_t``` just to validate that the setting is valid [1]

[1] http://tracker.ceph.com/issues/18465

dillaman commented:
retest this please

dillaman commented:
@trociny needs a rebase

trociny commented:
@dillaman updated

chenliuzhong commented:
@trociny   how to use "rbd_mirroring_replay_delay"?I try to use it by setting "rbd image-meta set pooltest/rbdtest conf_rbd_mirroring_replay_delay 600" for primary rbd image,but it did not work.Could you tell how to use this setting Correctly.
ceph version: 12.2.2
Thank you!

trociny commented:
Your image-meta command looks correct. I have just checked on the master and it works for me. How did you test that it did not work for you?

I was testing this way:

1) set conf_rbd_mirroring_replay_delay on the primary image
2) checked (with `rbd image-meta list`) on the secondary, that the setting was successfully mirrored
3) checked (with `rbd mirror image status`) on the secondary, that the status is replaying and  entries_behind_master is 0 (or small).
4) write some data to the primary image (with `rbd bench --io-type write`).
5) monitor `rbd mirror image status` on the scondary: entries_behind_master is  non zero, and only after 600 second it starts decreasing.

dillaman commented:
v12.2.2 has a bug where metadata config overrides are not applied to an already opened image (this is fixed in the forthcoming v12.2.3). Therefore, you could either set the "rbd_mirroring_replay_delay = XYZ" override in the rbd-mirror daemon's config file to globally apply a delay, wait for 12.2.3, or with 12.2.2 you can restart rbd-mirror after the metadata change propagates so that it picks up the new configuration option.

trociny commented:
Ah, I forgot about that bug.

@chenliuzhong also instead of restarting rbd-mirror daemon, you can just restart the image replayer via rbd-mirror admin socket:
```
ceph --admin-daemon  ${path_to_rbd_mirror_asok} rbd mirror restart ${POOL}/${image}
```

chenliuzhong commented:
I have try as step above:
(1)set conf_rbd_mirroring_replay_delay on the primary image  : ok
(2)checked (with rbd image-meta list) on the secondary, that the setting was successfully mirrored : ok
(3)checked (with rbd mirror image status) on the secondary, that the status is replaying and entries_behind_master is 0 (or small). : ok
(4) write some data to the primary image (with rbd bench --io-type write) :ok 
(5) monitor rbd mirror image status on the scondary: entries_behind_master is non zero, and only after 600 second it starts decreasing. :it  starts decrease just the first time not wait for 600 seconds.

I create rbd as : rbd create $RBD --size 1024 --pool $POOL --image-feature exclusive-lock,journaling
rbd info : 
``[root@node-1 ~]# rbd info pooltest/rbdtest
rbd image 'rbdtest':
	size 1024 MB in 256 objects
	order 22 (4096 kB objects)
	block_name_prefix: rbd_data.3744643c9869
	format: 2
	features: exclusive-lock, journaling
	flags: 
	create_timestamp: Wed Feb  7 02:55:07 2018
	journal: 3744643c9869
	mirroring state: enabled
	mirroring global id: b79761d2-58f8-4ef0-84fc-0d54e9c7640d
	mirroring primary: true
if I need to open other feature for this rbd.
Thank you!

chenliuzhong commented:
@dillaman  @trociny Thank you! is there a patch to solve  this issue now?

trociny commented:
You need these:

https://github.com/ceph/ceph/pull/19447
https://github.com/ceph/ceph/pull/19485

(ProTip! Add .patch or .diff to the end of URLs for Git’s plaintext views.)

chenliuzhong commented:
@trociny   OK ,thank you !
